File: Assets/Bundles/Server/Resources/TS/Services/Global/Shop/ShopService.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local ItemStack = require("Shared/TS/Inventory/ItemStack").ItemStack
local Network = require("Shared/TS/Network").Network
local DEFAULT_BEDWARS_SHOP = require("Shared/TS/Shop/ShopMeta").DEFAULT_BEDWARS_SHOP
local ShopService
do
	ShopService = setmetatable({}, {
		__tostring = function()
			return "ShopService"
		end,
	})
	ShopService.__index = ShopService
	function ShopService.new(...)
		local self = setmetatable({}, ShopService)
		return self:constructor(...) or self
	end
	function ShopService:constructor(entityService)
		self.entityService = entityService
	end
	function ShopService:OnStart()
		-- Handle incoming purchase requests.
		Network.ClientToServer.Shop.PurchaseRequest.Server:SetCallback(function(clientId, shopItem)
			return self:HandleIncomingPurchaseRequest(clientId, shopItem)
		end)
	end
	function ShopService:HandleIncomingPurchaseRequest(clientId, item)
		-- Validate that entity exists.
		local requestEntity = self.entityService:GetEntityByClientId(clientId)
		if not requestEntity then
			return false
		end
		-- Valide that entity has an inventory.
		if not (TS.instanceof(requestEntity, CharacterEntity)) then
			return false
		end
		-- Validate that shop item.
		local _shopItems = DEFAULT_BEDWARS_SHOP.shopItems
		local _arg0 = function(shopItem)
			return shopItem.item == item.item
		end
		-- ▼ ReadonlyArray.find ▼
		local _result
		for _i, _v in _shopItems do
			if _arg0(_v, _i - 1, _shopItems) == true then
				_result = _v
				break
			end
		end
		-- ▲ ReadonlyArray.find ▲
		local shopItem = _result
		if not shopItem then
			return false
		end
		-- Validate user can afford item.
		local entityInv = requestEntity:GetInventory()
		local canAfford = entityInv:HasEnough(shopItem.currency, shopItem.price)
		if not canAfford then
			return false
		end
		-- Fulfill purchase.
		entityInv:Decrement(shopItem.currency, shopItem.price)
		entityInv:AddItem(ItemStack.new(shopItem.item, shopItem.quantity))
		return true
	end
end
-- (Flamework) ShopService metadata
Reflect.defineMetadata(ShopService, "identifier", "Bundles/Server/Services/Global/Shop/ShopService@ShopService")
Reflect.defineMetadata(ShopService, "flamework:parameters", { "Bundles/Server/Services/Global/Entity/EntityService@EntityService" })
Reflect.defineMetadata(ShopService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ShopService, "$:flamework@Service", Service, { {} })
return {
	ShopService = ShopService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Global/Shop/ShopService.ts
import { OnStart, Service } from "@easy-games/flamework-core";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { ItemStack } from "Shared/Inventory/ItemStack";
import { Network } from "Shared/Network";
import { DEFAULT_BEDWARS_SHOP, ShopItem } from "Shared/Shop/ShopMeta";
import { EntityService } from "../Entity/EntityService";

@Service({})
export class ShopService implements OnStart {
	constructor(private readonly entityService: EntityService) {}

	OnStart(): void {
		/* Handle incoming purchase requests. */
		Network.ClientToServer.Shop.PurchaseRequest.Server.SetCallback((clientId, shopItem) => {
			return this.HandleIncomingPurchaseRequest(clientId, shopItem);
		});
	}

	/** Validates and fulfills incoming purchase requests. */
	private HandleIncomingPurchaseRequest(clientId: number, item: ShopItem): boolean {
		/* Validate that entity exists. */
		const requestEntity = this.entityService.GetEntityByClientId(clientId);
		if (!requestEntity) return false;
		/* Valide that entity has an inventory. */
		if (!(requestEntity instanceof CharacterEntity)) return false;
		/* Validate that shop item. */
		const shopItem = DEFAULT_BEDWARS_SHOP.shopItems.find((shopItem) => shopItem.itemType === item.itemType);
		if (!shopItem) return false;
		/* Validate user can afford item. */
		const entityInv = requestEntity.GetInventory();
		const canAfford = entityInv.HasEnough(shopItem.currency, shopItem.price);
		if (!canAfford) return false;
		/* Fulfill purchase. */
		entityInv.Decrement(shopItem.currency, shopItem.price);
		entityInv.AddItem(new ItemStack(shopItem.itemType, shopItem.quantity));
		return true;
	}
}
